src/components/Grid/types.ts

```diff
@@ -0,0 +1,170 @@
+  column: number;
+  event?: KeyEventType;
+}
+
+export interface RowSelectionAction {
+  type: "rowSelection" | "shiftRowSelection" | "ctrlRowSelection";
+  row: number;
+  event?: KeyEventType;
+}
+
+export type SelectionAction =
+  | AllSelection
+  | CellSelectionAction
+  | ColumnSelectionAction
+  | RowSelectionAction;
+
+export type SelectionType = "default" | "selectIndirect" | "selectDirect";
+
+export interface SelectionPos {
+  column: number;
+  row: number;
+}
+
+export interface EmptySelection {
+  type: "empty";
+}
+
+export interface RowsSelection {
+  type: "rows";
+  rows: Set<number>;
+  anchorRow: number;
+}
+
+export interface ColumnsSelection {
+  type: "columns";
+  columns: Set<number>;
+  anchorColumn: number;
+}
+
+export interface Rectangle {
+  left: number;
+  right: number;
+  top: number;
+  bottom: number;
+}
+
+export interface RectangleSelection {
+  type: "rectangle";
+  bounds: Rectangle;
+  anchor: SelectionPos;
+}
+
+export type SelectedRegion =
+  | EmptySelection
+  | RowsSelection
+  | ColumnsSelection
+  | RectangleSelection;
+
+export type RoundedType = "none" | "lg" | "md" | "sm";
+
+export type KeyEventType = "keypress" | "click";
+export type onSelectFn = (
+  action: SelectionAction,
+  selction: SelectedRegion,
+  focus: SelectionFocus
+) => void;
+
+type IsAllSelectedType = {
+  type: "all";
+};
+
+type IsRowSelectedType = {
+  type: "row";
+  row: number;
+};
+
+type IsColumnSelectedType = {
+  type: "column";
+  column: number;
+};
+
+type IsCellSelectedType = {
+  type: "cell";
+  row: number;
+  column: number;
+};
+
+export type IsSelectedType =
+  | IsAllSelectedType
+  | IsRowSelectedType
+  | IsColumnSelectedType
+  | IsCellSelectedType;
+
+export type SelectionTypeFn = (props: IsSelectedType) => SelectionType;
+
+export type ColumnResizeFn = (columnIndex: number, newWidth: number) => void;
+
+export type SelectionFocus = { row: number; column: number };
+
+export interface ItemDataType {
+  showRowNumber: boolean;
+  showHeader: boolean;
+  getSelectionType: SelectionTypeFn;
+  rowCount: number;
+  columnCount: number;
+  cell: CellProps;
+  focus: SelectionFocus;
+  rounded: RoundedType;
+  rowHeight: number;
+  headerHeight: number;
+  rowNumberWidth: number;
+}
+
+export interface GridContextMenuItemProps extends Omit<ContextMenuItemProps, "children"> {
+  label: ReactNode;
+}
+
+export interface GridProps
+  extends Omit<
+    VariableSizeGridProps,
+    | "height"
+    | "width"
+    | "rowHeight"
+    | "children"
+    | "innerElementType"
+    | "innerRef"
+    | "outerElementType"
+    | "outerRef"
+  > {
+  rowStart?: number;
+  rounded?: RoundedType;
+  focus: SelectionFocus;
+  rowHeight?: number;
+  cell: CellProps;
+  showHeader?: boolean;
+  showRowNumber?: boolean;
+  headerHeight?: number;
+  onFocusChange: (rowIndex: number, columnIndex: number) => void;
+  onSelect?: onSelectFn;
+  onColumnResize: ColumnResizeFn;
```

Comment:
this should have a default and not be expected from the user every time

Reply:
This is for the user to save in the state
Currently, we save it in sql-console in tab redux and when switching back it would be saved
Since the columnWidth is taken from the outside we use this function to update

Reply:
Giving this a default means you end up with a component that is sometimes controlled, sometimes uncontrolled depending on the properties. IME this is a recipe for over complicated state management code. I think a better approach is to have a component where everything is controlled, and then (if necessary) have wrapper components or hooks that provide simple defaults for simple use cases. I don't know whether such wrapper components would belong in click-ui or in the application repo itself though.